From c674bb5596460fddea5478208dc2cec17993189e Mon Sep 17 00:00:00 2001
From: "Richard W.M. Jones" <rjones@redhat.com>
Date: Wed, 13 Aug 2025 16:55:27 +0100
Subject: [PATCH] v2v: Fix SELinux relabelling

Update the common submodule to pick up:

  Richard W.M. Jones (3):
      mlcustomize/SELinux_relabel.ml: Add comment
      mlcustomize/SELinux_relabel.ml: Use new guestfs_setfiles API
      mlcustomize/SELinux_relabel.ml: Relabel every mountpoint

This allows SELinux relabelling to work for Linux guests that have a
split-/usr configuration.

This requires libguestfs >= 1.57.1, for the new guestfs_setfiles API.

Fixes: https://issues.redhat.com/browse/RHEL-108174
Reported-by: Germano Veit Michel
Thanks: Ming Xie
(cherry picked from commit e4c53263d08e8aa40b0d911d470ea2930dbb27c7)
---
 common                  | 2 +-
 m4/guestfs-libraries.m4 | 6 ++----
 2 files changed, 3 insertions(+), 5 deletions(-)

Submodule common 10d2b626..89f1eb2d:
diff --git a/common/mlcustomize/SELinux_relabel.ml b/common/mlcustomize/SELinux_relabel.ml
index 2f3a09bf..f1729e3f 100644
--- a/common/mlcustomize/SELinux_relabel.ml
+++ b/common/mlcustomize/SELinux_relabel.ml
@@ -1,5 +1,5 @@
 (* virt-customize
- * Copyright (C) 2016 Red Hat Inc.
+ * Copyright (C) 2016-2025 Red Hat Inc.
  *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
@@ -24,6 +24,10 @@ open Printf
 
 module G = Guestfs
 
+(* XXX A lot of this code could usefully be moved into
+ * [libguestfs.git/daemon/selinux.ml].
+ *)
+
 let rec relabel (g : G.guestfs) =
   (* Is the guest using SELinux?  (Otherwise this is a no-op). *)
   if is_selinux_guest g then (
@@ -109,5 +113,13 @@ and use_setfiles g =
     g#copy_attributes ~all:true old_specfile specfile
   );
 
+  (* Get the list of mountpoints, since setfiles does not cross
+   * filesystems (RHEL-108174).
+   *)
+  let mps = g#mountpoints () |>
+            List.map snd |>       (* the list of directories *)
+            List.sort compare |>  (* sort them for consistency *)
+            Array.of_list in
+
   (* Relabel everything. *)
-  g#selinux_relabel ~force:true specfile "/"
+  g#setfiles ~force:true specfile mps
diff --git a/m4/guestfs-libraries.m4 b/m4/guestfs-libraries.m4
index fa13e0f6..f1d5d127 100644
--- a/m4/guestfs-libraries.m4
+++ b/m4/guestfs-libraries.m4
@@ -19,10 +19,8 @@ dnl Any C libraries required by virt-v2v.
 
 dnl Of course we need libguestfs.
 dnl
-dnl We need libguestfs 1.55.6 for guestfs_sh_out.
-dnl We need libguestfs 1.55.12 for guestfs_btrfs_scrub_full.
-dnl We need libguestfs 1.55.13 for guestfs_e2fsck FORCENO flag.
-PKG_CHECK_MODULES([LIBGUESTFS], [libguestfs >= 1.55.13])
+dnl We need libguestfs 1.57.1 for guestfs_setfiles.
+PKG_CHECK_MODULES([LIBGUESTFS], [libguestfs >= 1.57.1])
 printf "libguestfs version is "; $PKG_CONFIG --modversion libguestfs
 
 dnl And libnbd.
